internal/fs/async_fetch_test.go

```diff
@@ -0,0 +1,161 @@
+const appendThreshold = 1
+
+type AsyncFetchTest struct {
+	ctx    context.Context
+	bucket gcsx.SyncerBucket
+	clock  timeutil.SimulatedClock
+	dh     *fs.DirHandle
+}
+
+var _ SetUpInterface = &AsyncFetchTest{}
+var _ TearDownInterface = &AsyncFetchTest{}
+
+func init() {
+	RegisterTestSuite(&AsyncFetchTest{})
+}
+
+func (t *AsyncFetchTest) SetUp(ti *TestInfo) {
+	t.ctx = ti.Ctx
+	t.clock.SetTime(time.Date(2015, 4, 5, 2, 15, 0, 0, time.Local))
+	bucket := gcsfake.NewFakeBucket(&t.clock, "some_bucket")
+	t.bucket = gcsx.NewSyncerBucket(
+		int64(appendThreshold), // Append threshold
+		tmpObjectPrefix,
+		bucket)
+}
+
+func TestAsync(t *testing.T) {
+	RunTests(t)
+}
+
+func (t *AsyncFetchTest) TearDown() {
+}
+
+// Create the inode and directory handle. No implicit dirs by default.
+func (t *AsyncFetchTest) createDirHandle(implicitDirs bool, enableNonexistentTypecache bool, dirInodeName string) {
+	in := inode.NewDirInode(
+		dirInodeID,
+		inode.NewDirName(inode.NewRootName(""), dirInodeName),
+		fuseops.InodeAttributes{
+			Uid:  uid,
+			Gid:  gid,
+			Mode: dirMode,
+		},
+		implicitDirs,
+		enableNonexistentTypecache,
+		typeCacheTTL,
+		&t.bucket,
+		&t.clock,
+		&t.clock)
+	t.dh = fs.NewDirHandle(in, implicitDirs)
+}
+
+func (t *AsyncFetchTest) resetDirHandle() {
+	t.dh = nil
+}
+
+// Directory Structure Used
+// foo   --Directory
+// fetchEntriesAsync will return 0 entries for empty directory
+func (t *AsyncFetchTest) FetchAsyncEntries_EmptyDir() {
+	t.createDirHandle(false, false, dirInodeName)
+	t.dh.FetchEntriesAsync(fuseops.RootInodeID, true)
+	AssertEq(len(t.dh.Entries), 0)
+	AssertEq(t.dh.EntriesValid, true)
+	t.resetDirHandle()
+}
+
+// Directory Structure Used
+// foo       --Directory
+// foo/bar   --File
+// fetchEntriesAsync will return 1 entry for directory with 1 file
+func (t *AsyncFetchTest) FetchAsyncEntries_NonEmptyDir() {
+	contents := "Non-empty dir"
+	filePath := path.Join(dirInodeName, fileUnderDir)
+	_, err := t.bucket.CreateObject(
+		t.ctx,
+		&gcs.CreateObjectRequest{
+			Name:     filePath,
+			Contents: strings.NewReader(contents),
+		})
+	AssertEq(err, nil)
+	t.createDirHandle(false, false, dirInodeName)
+	t.dh.FetchEntriesAsync(fuseops.RootInodeID, true)
+	AssertEq(len(t.dh.Entries), 1)
+	AssertEq(t.dh.Entries[0].Name, fileUnderDir)
+	AssertEq(t.dh.EntriesValid, true)
+	t.resetDirHandle()
+}
+
+// Directory Structure Used
+// foo              --Directory
+// foo/baz          --Implicit Directory
+// foo/baz/bar      --file
+// fetchEntriesAsync will return 1 entry for implicit directory if flag is set to true else 0
+func (t *AsyncFetchTest) FetchAsyncEntries_ImplicitDir() {
+	contents := "Implicit dir"
+	filePath := path.Join(dirInodeName, path.Join(implicitDirName, fileUnderDir))
+	_, err := t.bucket.CreateObject(
+		t.ctx,
+		&gcs.CreateObjectRequest{
+			Name:     filePath,
+			Contents: strings.NewReader(contents),
+		})
+	AssertEq(err, nil)
+
+	t.createDirHandle(true, false, dirInodeName)
```

Comment:
Add comment here --> something like implicit dir flag is true and same for below.